integration_tests/spec/spec_helper.rb

```diff
@@ -11,41 +11,98 @@
   config.filter_run focus: true
   config.run_all_when_everything_filtered = true
 
-  config.add_setting :mqtt
-  config.before(:suite) do
-    RSpec.configuration.mqtt = MQTT::Client.connect('mqtt://localhost')
-  end
+  # Set up an MQTT server
+  config.around(:each) do |example|
+    if ENV['CI_USE_EXTERNAL_MOSQUITTO'] == '1'
+      example.run
+      next
+    end
 
-  config.around do |example|
-    @inverter = Inverter.new
+    name = "mosquitto-lxp-bridge-tests-#{Process.pid}"
+    Kernel.system(
+      *%W[docker run --detach --rm -p 1883:1883 -p 9001:9001 --name #{name} eclipse-mosquitto:1.6],
+      out: '/dev/null'
+    )
+
+    begin
+      wait_for_socket('localhost', 1883)
+      example.run
+    ensure
+      Kernel.system(*%W[docker stop #{name}], out: '/dev/null')
+    end
+  end
 
+  # Set up an MQTT client, for expectations
+  config.around(:each) do |example|
+    @mqtt_client = MQTT::Client.connect('mqtt://localhost:1883')
     @messages = []
 
-    RSpec.configuration.mqtt.subscribe('#')
-    sub_thread = Thread.new do
-      RSpec.configuration.mqtt.get do |topic, data|
+    mqtt_client.subscribe('#')
+
+    thread = Thread.new do
+      mqtt_client.get do |topic, raw_data|
         data = begin
-          JSON.parse(data)
+          JSON.parse(raw_data)
         rescue JSON::ParserError
-          data
+          raw_data
         end
 
         @messages << { topic: topic, data: data }
       end
     end
 
-    example.run
+    begin
+      example.run
+    ensure
+      mqtt_client.disconnect
+      thread.kill
+      thread.join
+    end
+  end
+
+  config.around(:each) do |example|
+    @inverter = Inverter.new
+    inverter.run
+
+    begin
+      example.run
+    ensure
+      inverter.close
+    end
+  end
+
+  # Set up an lxp-bridge instance
+  config.before(:suite) do
+    Kernel.system(*%w[cargo build --manifest-path ../Cargo.toml])
```

Comment:
Ideally we'd run the integration test *after* the `build` step and have it use the produced binary, rather than needing all of rust to run the ruby integration suite.

Reply:
I am tempted to drop the build steps from CI that build native binaries, at least for PRs, and instead build a Docker image. Then we can use that for the integration tests?

The native builds are a bit of a relic from pre-docker days and I doubt anyone uses them to be honest. Could keep the binaries builds for releases as thats completely separate CI anyway but even then I think we should encourage Docker-only use.

Reply:
In this PR, I'm starting a new lxp-bridge instance per test. I *was* starting a new mosquitto per test, using docker, but that's weirdly broken inside CI, so I've stopped: https://github.com/celsworth/lxp-bridge/pull/149#discussion_r1141209493 . I expect an lxp-bridge docker container would have the same issue.

A docker image would just be the same binary sat inside a squashfs archive, with a few other gubbins; I wouldn't use docker for lx-bridge myself, but I don't much mind how CI works - as long as it does work :sweat_smile:.